core/lib/types/src/fee_model.rs

```diff
@@ -0,0 +1,179 @@
+    pub fn into_l1_pegged(self) -> L1PeggedBatchFeeModelInput {
+        match self {
+            BatchFeeInput::L1Pegged(input) => input,
+            _ => panic!("Can not convert PubdataIndependentBatchFeeModelInput into L1PeggedBatchFeeModelInput"),
+        }
+    }
+
+    pub fn l1_pegged_ref_mut(&mut self) -> &mut L1PeggedBatchFeeModelInput {
+        match self {
+            BatchFeeInput::L1Pegged(input) => input,
+            _ => panic!("Can not convert PubdataIndependentBatchFeeModelInput into L1PeggedBatchFeeModelInput"),
+        }
+    }
+
+    pub fn l1_pegged_ref(&self) -> &L1PeggedBatchFeeModelInput {
+        match self {
+            BatchFeeInput::L1Pegged(input) => input,
+            _ => panic!("Can not convert PubdataIndependentBatchFeeModelInput into L1PeggedBatchFeeModelInput"),
+        }
+    }
+
+    pub fn fair_pubdata_price(&self) -> u64 {
+        match self {
+            BatchFeeInput::L1Pegged(input) => input.l1_gas_price * L1_GAS_PER_PUBDATA_BYTE as u64,
+            BatchFeeInput::PubdataIndependent(input) => input.fair_pubdata_price,
+        }
+    }
+
+    pub fn fair_l2_gas_price(&self) -> u64 {
+        match self {
+            BatchFeeInput::L1Pegged(input) => input.fair_l2_gas_price,
+            BatchFeeInput::PubdataIndependent(input) => input.fair_l2_gas_price,
+        }
+    }
+
+    pub fn l1_gas_price(&self) -> u64 {
+        match self {
+            BatchFeeInput::L1Pegged(input) => input.l1_gas_price,
+            BatchFeeInput::PubdataIndependent(input) => input.l1_gas_price,
+        }
+    }
+
+    pub fn into_pubdata_independent(self) -> PubdataIndependentBatchFeeModelInput {
+        match self {
+            BatchFeeInput::PubdataIndependent(input) => input,
+            BatchFeeInput::L1Pegged(input) => PubdataIndependentBatchFeeModelInput {
+                fair_l2_gas_price: input.fair_l2_gas_price,
+                fair_pubdata_price: input.l1_gas_price * 17,
+                l1_gas_price: input.l1_gas_price,
+            },
+        }
+    }
+
+    pub fn for_protocol_version(
+        protocol_version: ProtocolVersionId,
+        fair_l2_gas_price: u64,
+        fair_pubdata_price: Option<u64>,
+        l1_gas_price: u64,
+    ) -> Self {
+        if protocol_version.is_1_4_1() {
+            Self::PubdataIndependent(PubdataIndependentBatchFeeModelInput {
+                fair_l2_gas_price,
+                fair_pubdata_price: fair_pubdata_price
+                    .expect("Pubdata price must be provided for protocol version 1.4.1"),
+                l1_gas_price,
+            })
+        } else {
+            Self::L1Pegged(L1PeggedBatchFeeModelInput {
+                fair_l2_gas_price,
+                l1_gas_price,
+            })
+        }
+    }
+}
+
+/// Pubdata is only published via calldata and so its price is pegged to the L1 gas price.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct L1PeggedBatchFeeModelInput {
+    /// Fair L2 gas price to provide
```

Comment:
we should explain more what 'fair' here means. I guess it should 'only' cover the CPU / prover costs, right?

Reply:
No, actually "fair" would need to cover the L1 batch overhead costs in the new fee model